docs: #863 second high availability blog (#1124)
redpanda-data/documentation: docs/deploy/deployment-option/self-hosted/manual/high-availability.mdx

|  Failure     | Impact | Mitigation strategy |
| ----------- | ----------- | ----------- |
| Node failure | Loss of function for an individual node or for any virtual machine (VM) hosted on that node | Multi-node cluster |
| Rack or switch failure | Loss of nodes/VMs hosted within that rack, or loss of connectivity to them | Multi-node cluster spread across multiple racks or network failure domains |
| Data center failure | Loss of nodes/VMs hosted within that data center, or loss of connectivity to them | Multi-AZ cluster or replicated deployment |
| Region failure | Loss of nodes/VMs hosted within that region, or loss of connectivity to them | Geo-stretch cluster (latency dependent) or replicated deployment |
| Global, systemic outage (DNS failures, routing failures) | Complete outage for all systems and services impacting customers and staff | Offline backups (Tiered Storage), replicas in 3rd-party domains |
| Data loss or corruption (accidental or malicious) | Corrupt or unavailable data that also affects synchronous replicas | Offline backups (Tiered Storage) |

## HA deployments

> Mission-critical deployments typically employ the following strategies:

Author: added back an intro line

- [MirrorMaker2 replication](../../../../../manage/data-migration)
- [Remote Read Replicas](../../../../../manage/remote-read-replicas)
- [Redpanda Edge Agent](https://github.com/redpanda-data/redpanda-edge-agent)

Alternatively, you could dual-feed clusters in multiple regions. Dual feeding is the process of having producers connect to your cluster across multiple regions. However, this introduces additional complexity onto the producing application. It also requires consumers that have sufficient deduplication logic built in to handle offsets, since they won’t be the same across each cluster.

### Deployment strategies

Consider the following Redpanda deployment strategies for the most common types of failures. 

|  Failure     | Impact | Mitigation strategy |
| ----------- | ----------- | ----------- |
| Node failure | Loss of function for an individual node or for any virtual machine (VM) hosted on that node | Multi-node deployment |

> | Rack or switch failure | Loss of nodes/VMs hosted within that rack, or loss of connectivity to them | Multi-node deployment spread across multiple racks or network failure domains |

Author: @vuldin, @tmgstevens: Reminder to please review this updated HA page! And please see Jake's comment about whether VM is the correct term we should be using on this page.

> | Rack or switch failure | Loss of nodes/VMs hosted within that rack, or loss of connectivity to them | Multi-node deployment spread across multiple racks or network failure domains |

Author: > We should clarify whether VM is the correct term for the machine that Redpanda runs on. It may not always be a VM.

@tmgstevens This HA page uses VM throughout. Would you please confirm that that's the correct term to use?

> | Rack or switch failure | Loss of nodes/VMs hosted within that rack, or loss of connectivity to them | Multi-node deployment spread across multiple racks or network failure domains |

Author: > We should clarify whether VM is the correct term for the machine that Redpanda runs on. It may not always be a VM.

@vuldin I think this is the last open question on this page. Can you please check this and give the page a final look before approving?

> | Rack or switch failure | Loss of nodes/VMs hosted within that rack, or loss of connectivity to them | Multi-node deployment spread across multiple racks or network failure domains |

Reviewer: I think this should stay as node

> While Apache Kafka® uses in-sync replicas, Redpanda uses a quorum-based majority with the Raft replication protocol. Kafka performance is negatively impacted when any replica is out-of-sync, but a Redpanda cluster tolerates replica failures without any performance degradation.  

Reviewer: Kafka performance is negative impacted when any "in-sync" replica is running slower than other replicas in the ISR set.

Author: fixed

> Monitor the health of your cluster with the [`rpk cluster health`](../../../../../reference/rpk/rpk-cluster/rpk-cluster-health) command. This tells you how many brokers are down (if any), and if you have any leaderless partitions. 

Reviewer 2: ```suggestion
Monitor the health of your cluster with the [`rpk cluster health`](../../../../../reference/rpk/rpk-cluster/rpk-cluster-health) command, which tells you how many brokers are down (if any), and if you have any leaderless partitions. 
```

> Monitor the health of your cluster with the [`rpk cluster health`](../../../../../reference/rpk/rpk-cluster/rpk-cluster-health) command. This tells you how many brokers are down (if any), and if you have any leaderless partitions. 

Reviewer: ```suggestion
Monitor the health of your cluster with the [`rpk cluster health`](../../../../../reference/rpk/rpk-cluster/rpk-cluster-health) command. This tells you if any brokers are down, and if you have any leaderless partitions. 
```

2. The replicas receive messages written to the partition of the leader. They send acknowledgments to the leader after successfully replicating the message to their internal partition. 
3. The leader sends an acknowledgment to the producer of the message, as determined by that producer’s `acks` value. Redpanda considers the group consistent after a majority has formed consensus; that is, a majority of participants acknowledged the write.

> While Apache Kafka® uses in-sync replicas, Redpanda uses a quorum-based majority with the Raft replication protocol. Kafka performance is negative impacted when any "in-sync" replica is running slower than other replicas in the In-Sync Replica (ISR) set.

Author: Yes, thank you--I'll fix it!

  ```

2. Clone the deployment-automation repository:
   
  ```bash
  $ git clone https://github.com/redpanda-data/deployment-automation
  ```

1. Initialize Terraform for your cloud provider:

  ```bash
  $ cd deployment-automation/aws (or cd deployment-automation/azure, or cd deployment-automation/gcp)
  $ terraform init
    ```

4. Deploy the infrastructure (this assumes you have cloud credentials available):

  ```bash
  $ terraform apply -var=ha=true
    ```

5. Verify that the racks have been correctly specified in the `host.ini` file:

  ```bash
  $ cd ..
  $ cat hosts.ini
  ```

  <details>
  <summary>
  Example output
  </summary>

  ```
  [redpanda]
  35.166.210.85 ansible_user=ubuntu ansible_become=True private_ip=172.31.7.173 id=0 rack=1
  18.237.173.220 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.138 id=1 rack=2
  54.218.103.91 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.93 id=2 rack=3
  …
  ```

  </details>

6. Provision the cluster with Ansible:

  ```bash
  $ ansible-playbook --private-key <your_private_key> ansible/playbooks/provision-node.yml -i hosts.ini
  ```

7. Verify that rack awareness is enabled:
   

>   a.  Get Redpanda host names:

Author: @JakeSCahill: Nothing in our style guide yet, but at Oracle our guidance for host name and user name was to use two words (just hostname and username as a one-word variable in code instructions and commands). Do you feel strong about usage as word in text?

>   a.  Get Redpanda host names:

Reviewer 3: I don't feel strongly, tbh. If you do an internal search in our docs for hostname and username, you seem to get more results. Personally, I see hostname and username used more widely in docs/applications. But maybe it's something we should add to our next team meeting agenda to discuss 😄

>   a.  Get Redpanda host names:

Author: Thx, I added to Friday's agenda!

>   a.  Get Redpanda host names:

Author: adding hostname and username to our style guide as one word

    
    ```bash
    $ grep -A1 '\[redpanda]' hosts.ini
    ``` 

    <details>
    <summary>
    Example output
    </summary>

    ```
    20.80.113.122 ansible_user=adminpanda ansible_become=True private_ip=10.0.1.7 rack=0
    ```

    </details>

>   b. SSH into a cluster node with the user name and host name of that Redpanda node:

Reviewer: I agree - I think host is a better term for this.

>   b. SSH into a cluster node with the user name and host name of that Redpanda node:

Author: > I agree - I think host is a better term for this.

@tmgstevens please confirm:

See also: [Tiered Storage](../../../../../manage/tiered-storage)

## Single-AZ deployment details

> When deploying a cluster in a single AZ, you need to ensure that, within the AZ, there are no single points of failure and that Redpanda spreads its replicas across the failure domains. This is achieved with Redpanda’s rack awareness, which deploys *n* Redpanda brokers across three or more racks, or failure domains, within the AZ.

Reviewer: ```suggestion
When deploying a cluster for High Availablity into a single AZ or Data Center, you need to ensure that, within the AZ, single points of failure are minimized and that Redpanda is configured to be aware of any discrete failure domains within the AZ. This is achieved with Redpanda’s rack awareness, which deploys *n* Redpanda brokers across three or more racks (or failure domains) within the AZ.
```

  ```bash
  cd ..
  cat hosts.ini
  ```

  <details>
  <summary>
  Example output
  </summary>

  ```
  [redpanda]
  35.166.210.85 ansible_user=ubuntu ansible_become=True private_ip=172.31.7.173 id=0 rack=1
  18.237.173.220 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.138 id=1 rack=2
  54.218.103.91 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.93 id=2 rack=3
  …
  ```

  </details>

6. Provision the cluster with Ansible:

  ```bash
  ansible-playbook --private-key <your_private_key> ansible/playbooks/provision-node.yml -i hosts.ini
  ```

7. Verify that rack awareness is enabled:
   
  a.  Get Redpanda hostnames:
    
    ```bash
    grep -A1 '\[redpanda]' hosts.ini
    ``` 

    <details>
    <summary>
    Example output
    </summary>

    ```
    20.80.113.122 ansible_user=adminpanda ansible_become=True private_ip=10.0.1.7 rack=0
    ```

    </details>

  b. SSH into a cluster broker with the username and hostname of that Redpanda broker:

    ```bash
    ssh -i ~/.ssh/id_rsa <username>@<hostname of redpanda broker>
    ```

  c. Verify that rack awareness is enabled:

    ```bash
    rpk cluster config get enable_rack_awareness
    true
    ```

>   d. Check the rack assigned to this specific broker:

Reviewer 2: Check it for what? Can you clarify? Is this "Verify the rack is assigned to the specific broker"?

Reviewer 4: I just verified that rack IDs should match as you mention and provided a suggestion in my review. Thanks for catching this!

    ```bash
    rpk cluster status

>     CLUSTER

Reviewer: This is example output as well

Author: fixed

### Configure rack awareness

To make Redpanda aware of the topology it's running on, you need to configure the cluster to enable rack awareness and then configure each broker with the identifier of the rack.

Set the `enable_rack_awareness custer` property either in `/etc/redpanda/.bootstrap.yaml` or with `rpk`:

```bash
rpk cluster config set enable_rack_awareness true
```

For each broker, set the rack ID in `/etc/redpanda/redpanda.yaml` file or with `rpk`:

```bash
rpk redpanda config set redpanda.rack <rackid>
```

The modified Ansible playbooks take a per-instance rack variable from the Terraform output and use that to set the relevant cluster and broker configuration. Redpanda deployment automation can provision public cloud infrastructure with discrete failure domains (`-var=ha=true`) and use the resulting inventory to provision rack-aware clusters using Ansible.

With rack awareness enabled, Redpanda ensures that no more than a minority of replicas appear on a single rack, even during cluster balancing.

> ### Single-AZ Example

Reviewer 4: I feel like this section should be it's own page, and then the rest of the page would be focused on primarily definition and explanation. This section focuses on creating a single-AZ deployment and is very prescriptive, unlike the rest of the page. Maybe instead of including a single-AZ example, this could be an Examples section that links to a single-AZ page (and, in the future, other examples as we add them)?

The following example deploys an HA cluster into AWS, Azure, or GCP using Terraform and Ansible.

1. Initialize a private key, if you haven’t done so already:

  ```bash
  ssh-keygen -f ~/.ssh/id_rsa
  ```

2. Clone the deployment-automation repository:
   
  ```bash
  git clone https://github.com/redpanda-data/deployment-automation
  ```

> 

Reviewer 4: There needs to another step here, which is to take a look at the pre-requisites and install the ansible requirements. Since suggestions within a PR comment doesn't work for code snippets, I'll just include the new section below:

===

1. Make sure you install the pre-requisites listed [here](https://github.com/redpanda-data/deployment-automation#installation-prerequisites). This includes installing ansible requirements with the following command:

```bash
ansible-galaxy install -r ansible/requirements.yml
```

Author: Thanks, I've added this.

1. Initialize Terraform for your cloud provider:

  ```bash
  cd deployment-automation/aws (or cd deployment-automation/azure, or cd deployment-automation/gcp)
  terraform init
    ```

4. Deploy the infrastructure (this assumes you have cloud credentials available):

  ```bash
  terraform apply -var=ha=true
    ```

5. Verify that the racks have been correctly specified in the `host.ini` file:

  ```bash
  cd ..
  cat hosts.ini
  ```

  <details>
  <summary>
  Example output
  </summary>

  ```
  [redpanda]

>   35.166.210.85 ansible_user=ubuntu ansible_become=True private_ip=172.31.7.173 id=0 rack=1
>   18.237.173.220 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.138 id=1 rack=2
>   54.218.103.91 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.93 id=2 rack=3

Reviewer 4: ```suggestion
  35.166.210.85 ansible_user=ubuntu ansible_become=True private_ip=172.31.7.173 rack=1
  18.237.173.220 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.138 rack=2
  54.218.103.91 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.93 rack=3
```

  …
  ```

  </details>

6. Provision the cluster with Ansible:

  ```bash

>   ansible-playbook --private-key <your_private_key> ansible/playbooks/provision-node.yml -i hosts.ini

Reviewer 4: ```suggestion
  ansible-playbook --private-key `cat ~/.ssh/id_rsa.pub | awk '{print $2}'` ansible/playbooks/provision-node.yml -i hosts.ini
```

  ```

7. Verify that rack awareness is enabled:
   
  a.  Get Redpanda hostnames:
    
    ```bash
    grep -A1 '\[redpanda]' hosts.ini
    ``` 

    <details>
    <summary>
    Example output
    </summary>

    ```
    20.80.113.122 ansible_user=adminpanda ansible_become=True private_ip=10.0.1.7 rack=0
    ```

    </details>

  b. SSH into a cluster broker with the username and hostname of that Redpanda broker:

    ```bash
    ssh -i ~/.ssh/id_rsa <username>@<hostname of redpanda broker>
    ```

  c. Verify that rack awareness is enabled:

    ```bash
    rpk cluster config get enable_rack_awareness
    ```
    <details>
    <summary>
    Example output
    </summary>

    ```
    true
    ```

    </details>

  d. Check the rack assigned to this specific broker:

    ```bash
    rpk cluster status
    ```
    <details>
    <summary>
    Example output
    </summary>

    ```
    CLUSTER
    = = = =
    redpanda.807d59af-e033-466a-98c3-bb0be15c255d

    BROKERS
    = = = =
    ID HOST PORT RACK

>     0* 10.0.1.7 9092 0
>     1 10.0.1.4 9092 1
>     2 10.0.1.8 9092 2

Reviewer 4: ```suggestion
    0* 10.0.1.7 9092 1
    1 10.0.1.4 9092 2
    2 10.0.1.8 9092 3
```
